crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -3270,33 +3219,104 @@ impl<TX: DbTxMut + DbTx + 'static, N: ProviderNodeTypes + 'static> BlockWriter
             "Inserted block"
         );
 
-        Ok(block_indices)
+        Ok(StoredBlockBodyIndices { first_tx_num, tx_count })
     }
 
     fn append_block_bodies(
         &self,
         bodies: Vec<(BlockNumber, Option<Self::Body>)>,
+        write_transactions_to: StorageLocation,
     ) -> ProviderResult<()> {
+        let Some(from_block) = bodies.first().map(|(block, _)| *block) else { return Ok(()) };
+
+        // Initialize writer if we will be writing transactions to staticfiles
+        let mut tx_static_writer = write_transactions_to
+            .static_files()
+            .then(|| {
+                self.static_file_provider.get_writer(from_block, StaticFileSegment::Transactions)
+            })
+            .transpose()?;
+
         let mut block_indices_cursor = self.tx.cursor_write::<tables::BlockBodyIndices>()?;
         let mut tx_block_cursor = self.tx.cursor_write::<tables::TransactionBlocks>()?;
 
+        // Initialize cursor if we will be writing transactions to database
+        let mut tx_cursor = write_transactions_to
+            .database()
+            .then(|| {
+                self.tx.cursor_write::<tables::Transactions<
+                    <Self::Body as reth_primitives_traits::BlockBody>::Transaction,
+                >>()
+            })
+            .transpose()?;
+
         // Get id for the next tx_num of zero if there are no transactions.
         let mut next_tx_num = tx_block_cursor.last()?.map(|(id, _)| id + 1).unwrap_or_default();
 
         for (block_number, body) in &bodies {
+            // Increment block on static file header.
+            if let Some(writer) = tx_static_writer.as_mut() {
+                writer.increment_block(*block_number)?;
+            }
+
             let tx_count = body.as_ref().map(|b| b.transactions().len() as u64).unwrap_or_default();
             let block_indices = StoredBlockBodyIndices { first_tx_num: next_tx_num, tx_count };
 
+            let mut durations_recorder = metrics::DurationsRecorder::default();
+
             // insert block meta
             block_indices_cursor.append(*block_number, block_indices)?;
 
-            next_tx_num += tx_count;
+            durations_recorder.record_relative(metrics::Action::InsertBlockBodyIndices);
+
             let Some(body) = body else { continue };
 
             // write transaction block index
             if !body.transactions().is_empty() {
                 tx_block_cursor.append(block_indices.last_tx_num(), *block_number)?;
+                durations_recorder.record_relative(metrics::Action::InsertTransactionBlocks);
+            }
+
+            let mut transactions_elapsed = Duration::default();
+
+            // write transactions
+            for transaction in body.transactions() {
+                let hash = transaction.tx_hash();
+                let start = Instant::now();
```

Comment:
this is way too expensive to do per tx

was this the case before?

I'd rather not do this at all

Reply:
this was the case for live sync but not for the pipeline, should I remove it or just make optional?

Reply:
don't want to block this pr for this,
but would like to get rid of this detailed time tracking entirely because these additional syscall per tx are likely expensive (400 timesyscalls for a 200txs block)